api/src/main/java/marquez/db/JobVersionDao.java

```diff
@@ -14,205 +14,303 @@
+  /**
+   * Returns the input or output datasets for a given job version.
+   *
+   * @param jobVersionUuid The unique ID of the job version.
+   * @param ioType The {@link IoType} of the dataset.
+   */
+  @SqlQuery(
+      "SELECT dataset_uuid FROM job_versions_io_mapping "
+          + "WHERE job_version_uuid = :jobVersionUuid AND io_type = :ioType")
+  List<UUID> findInputOrOutputDatasetsFor(UUID jobVersionUuid, IoType ioType);
+
+  /**
+   * Used to associate a {@link Run} to a given job version. A run is an instance of a job version.
+   * When a run object is instantiated, the {@code latest_run_uuid} column in the {@code
+   * job_versions} table is updated and set to the unique ID of the latest run of the version. Note,
+   * multiple run instances may be linked to a job version as runs are based on a version.
+   *
+   * @param jobVersionUuid The unique ID of the job version.
+   * @param updatedAt The last modified timestamp of the job version.
+   * @param latestRunUuid The unique ID of the {@link Run} associated with the job version.
+   */
+  @SqlUpdate(
+      "UPDATE job_versions "
+          + "SET updated_at = :updatedAt, "
+          + "    latest_run_uuid = :latestRunUuid "
+          + "WHERE uuid = :jobVersionUuid")
+  void updateLatestRunFor(UUID jobVersionUuid, Instant updatedAt, UUID latestRunUuid);
+
+  /** Returns the unique ID of the latest {@link Run} for a given job version. */
+  @SqlQuery("SELECT latest_run_uuid FROM job_versions WHERE uuid = :jobVersionUuid")
+  Optional<UUID> findLatestRunFor(UUID jobVersionUuid);
+
+  /** Returns the total row count for the {@code job_versions} table; used for testing only. */
+  @VisibleForTesting
+  @SqlQuery("SELECT COUNT(*) FROM job_versions")
+  int count();
+
+  /**
+   * Used to upsert an immutable {@link JobVersionRow} object when a {@link Run} has transitioned. A
+   * {@link Version} is generated using {@link Utils#newJobVersionFor(String, String, List, List,
+   * Map, String)} based on the jobs inputs and inputs, source code location, and context. A version
+   * for a given job is created <i>only</i> when a {@link Run} transitions into a {@code COMPLETED},
+   * {@code ABORTED}, or {@code FAILED} state.
+   *
+   * @param namespaceName The namespace for the job version.
+   * @param jobName The name of the job.
+   * @param runUuid The unique ID of the run associated with the job version.
+   * @param runState The current run state.
+   * @param transitionedAt The timestamp of the run state transition.
+   * @return A {@link BagOfJobVersionInfo} object.
+   */
+  @Transaction
+  default BagOfJobVersionInfo upsertJobVersionOnRunTransition(
+      @NonNull String namespaceName,
+      @NonNull String jobName,
+      @NonNull UUID runUuid,
+      @NonNull RunState runState,
+      @NonNull Instant transitionedAt) {
+    // Get the job.
+    final JobDao jobDao = createJobDao();
+    final JobRow jobRow = jobDao.findJobByNameAsRow(namespaceName, jobName).get();
```

Comment:
Shouldn't we handle if the Optional is empty?

Reply:
We should, but most of the DAO classes handle optionals this way (not a great practive). I've opened an issue to address this behavior across classes: https://github.com/MarquezProject/marquez/issues/1276